core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,360 @@
+
+    @GuardedBy("this")
+    private final long[] writerRowCountSinceLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionWrittenBytesAtLastRebalance;
+
+    public UniformPartitionRebalancer(
+            List<Supplier<Long>> writerPhysicalWrittenBytesSuppliers,
+            int partitionCount,
+            int numberOfWriters,
+            long writerMinSize)
+    {
+        this.writerPhysicalWrittenBytesSuppliers = requireNonNull(writerPhysicalWrittenBytesSuppliers, "writerPhysicalWrittenBytesSuppliers is null");
+        this.writerMinSize = writerMinSize;
+        this.numberOfWriters = numberOfWriters;
+
+        this.writerPhysicalWrittenBytesAtLastRebalance = new AtomicLongArray(numberOfWriters);
+        this.writerWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+        this.writerRowCountSinceLastRebalance = new long[numberOfWriters];
+        this.partitionWrittenBytesAtLastRebalance = new long[partitionCount];
+
+        List<List<Integer>> partitionToWritersAssignment = new ArrayList<>();
+        for (int i = 0; i < partitionCount; i++) {
+            partitionToWritersAssignment.add(new CopyOnWriteArrayList<>(ImmutableList.of(i % numberOfWriters)));
+        }
+        this.partitionToWritersAssignment = new AtomicReference<>(partitionToWritersAssignment);
+    }
+
+    public List<Integer> getWriterIds(int partitionId)
+    {
+        return partitionToWritersAssignment.get().get(partitionId);
+    }
+
+    public synchronized void addPartitionsRowCount(Map<WriterPartitionId, Integer> partitionsRowCount)
+    {
+        partitionsRowCount.forEach((writerPartitionId, rowCount) -> {
+            writerRowCountSinceLastRebalance[writerPartitionId.writerId] += rowCount;
+            writerPartitionIdToRowCount.compute(writerPartitionId, (key, currentRowCount) -> {
+                if (currentRowCount == null) {
+                    return (long) rowCount;
+                }
+                return currentRowCount + rowCount;
+            });
+        });
+    }
+
+    public void rebalancePartitions()
+    {
+        List<Long> writerPhysicalWrittenBytes = writerPhysicalWrittenBytesSuppliers.stream()
+                .map(Supplier::get)
+                .collect(toImmutableList());
+
+        // Rebalance only when total bytes written since last rebalance is greater than MIN_PHYSICAL_WRITTEN_REBALANCE_BYTES (100 MB)
+        if (getPhysicalWrittenBytesSinceLastRebalance(writerPhysicalWrittenBytes) > MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES) {
+            rebalancePartitions(writerPhysicalWrittenBytes);
+        }
+    }
+
+    private synchronized void rebalancePartitions(List<Long> writerPhysicalWrittenBytes)
+    {
+        // Rebalance only when total bytes written since last rebalance is greater than MIN_PHYSICAL_WRITTEN_REBALANCE_BYTES (100 MB)
+        if (getPhysicalWrittenBytesSinceLastRebalance(writerPhysicalWrittenBytes) <= MIN_REBALANCE_PHYSICAL_WRITTEN_BYTES) {
```

Comment:
This was already checked above

Reply:
Yeah but that is outside synchronized. It's possible that multiple threads calling the above method at the same time. Therefore, we need to have this check in the critical section too.